test/construction.test.js

```diff
@@ -355,17 +355,146 @@ describe('construction', function() {
         ['number', 'string', 'boolean']);
   });
 
-  it('should allow a function to be defined recursively', function () {
+  it('should allow a function refer to itself', function () {
     var fn = typed({
       'number': function (value) {
         return 'number:' + value;
       },
-      'string': function (value) {
-        return this(parseInt(value, 10));
-      }
+      'string': typed.referToSelf((self) => {
+        return function (value) {
+          assert.strictEqual(self, fn)
+
+          return self(parseInt(value, 10));
+        }
+      })
+    });
+
+    assert.equal(fn('2'), 'number:2');
+  });
+
+  it('should allow to resolve function signatures with referTo', function () {
+    var fnNumber = function (value) {
+      return 'number:' + value;
+    }
+
+    var fnBoolean = function (value) {
+      return 'boolean:' + value;
+    }
+
+    var fn = typed({
+      'number': fnNumber,
+      'boolean': fnBoolean,
+      'string': typed.referTo('number', 'boolean', (fnNumberResolved, fnBooleanResolved) => {
+        assert.strictEqual(fnNumberResolved, fnNumber)
+        assert.strictEqual(fnBooleanResolved, fnBoolean)
+
+        return function fnString(value) {
+          return fnNumberResolved(parseInt(value, 10));
+        }
+      })
     });
 
     assert.equal(fn('2'), 'number:2');
+  });
+
+  it('should throw an exception when a signature is not found with referTo', function () {
+    assert.throws(() => {
+      typed({
+        'string': typed.referTo('number', (fnNumberResolved) => {
+          return function fnString(value) {
+            return fnNumberResolved(parseInt(value, 10));
+          }
+        })
+      });
+    }, /Cannot refer to signature "string": reference signature "number" not found/)
+  })
+
+  it('should throw an exception when a signature is not resolved with referTo', function () {
+    assert.throws(() => {
+      typed({
+        'string': typed.referTo('number', (fnNumberResolved) => {
+          return function fnString(value) {
+            return fnNumberResolved(parseInt(value, 10));
+          }
+        }),
+        // we define `number` after we use it in `string` to enforce getting an error
+        'number': typed.referTo(() => {
+          return 'number:' + value;
+        })
+      });
+    }, /Cannot refer to signature "string": signature is referring to a signature "number" which is not yet resolved/)
+  })
+
```

Comment:
Yes good point. I've been thinking shortly about this. You're right that this can will put the burden on the user, having to make sure there are no forward references. For a single function it's not a big problem I think, but it can get nasty when trying to merge two functions together. 

What I like about the current implementation though is that it is functional and very simple and predictable.

I'm not sure but it may turn out a relatively complex to do this auto-dependency-resolving neatly: you can get a chain of dependencies, you can get circular dependencies, etc. If OK with you I prefer to address that in a separate PR, and only as soon as the need arises. The current implementation is functional and we can improve on it in a later stage making it smarter and more convenient. At this moment I have the feeling that needing `referTo` is already a bit of an edge case, so I'm not sure if it will be a common case to have a typed-function with "lots" of signatures referring to each other. If the implementation for auto resolving these dependencies is relatively complex, I'm not sure if it is worth it. So I want to be a bit cautious in that regard spending time and increasing code complexity.

Reply:
> At this moment I have the feeling that needing referTo is already a bit of an edge case, so I'm not sure if it will be a common case to have a typed-function with "lots" of signatures referring to each other.

I am confused; a good fraction of mathjs typed-functions use this, certainly dozens of them, and simplify() has numerous forward references, and derivative() has double references (implementation A refers to B refers to C) although those happen to all be arranged (so far as I can tell, it's a bit tricky) so all of the references are back references. I didn't keep searching for a double forward reference, but I would not be surprised if they are there. Adoption of v3 in mathjs is already going to be a big change (although well worth it), which means it would be nice if the changing of this --> referTo could be done as an editor macro, rather than having to worry about re-ordering the implementations of a typed-function as well.

As for the code to handle forward references, it's very simple: you just take the current resolution loop, and rather than erroring when unresolved, you set a flag, and also you set a flag when you perform a resolution. Then you just repeat the loop until the first time you perform no resolutions, and if the flag that there were unresolved references is set, you issue a "circular reference" error (you can use the item that was not resolved as the "true" value of the flag so that the error message can tell you where the circular reference occurred), and if there were no unresolved references, you're good to go.

I would be happy to add this to this PR if you would like; I really think we will save ourselves a significant amount of trouble with the mathjs adoption of v3 if we do.

Reply:
P.S. I think we mean different things by the word "functional," I don't see anything more or less functional about supporting forward references or not supporting them.

Reply:
> I am confused; a good fraction of mathjs typed-functions use this, certainly dozens of them, and simplify() has numerous forward references

o wow, you're right. In my mind it wasn't used that much in mathjs but I just had a look at the code, there are 184 occurrences of `this(...)`. Thanks for correcting me. That will be quite some refactoring indeed. I hope we can use `referTo` a lot, that will improve runtime performance.

> As for the code to handle forward references, it's very simple: you just take the current resolution loop, and rather than erroring when unresolved, you set a flag, and also you set a flag when you perform a resolution.

It's probably not that complex as I first thought. Since `resolveReferences` still need a refactor, it can well be that it turn out to be the simplest solution is a recursive one which directly solves the forward references and circular dependencies.

> I would be happy to add this to this PR if you would like; I really think we will save ourselves a significant amount of trouble with the mathjs adoption of v3 if we do.

Thanks, a bit of help on this would be welcome! I'll make you a collaborator on `typed-function` too so you can directly work on this `feat/no_binding_self2` feature branch. I don't expect to start on this refactor today at least, but will let you know if I happen too after all.

> P.S. I think we mean different things by the word "functional," I don't see anything more or less functional about supporting forward references or not supporting them.

😂  that is a really nice one. I hadn't realized my confusing wording of _"it is functional and very simple"_, what I just meant there is _"it works and is very simple"_. Fits well though on Aprils Fools' Day 😅

Reply:
> Thanks, a bit of help on this would be welcome!

OK, I pushed a branch `feat/no_binding_self_refactor` to this repo. I didn't know if you'd want me to make a pull request to this branch or if you can just look at that branch and see if it looks good and switch to it if so, or what would be the most helpful, but I definitely thought you'd want to compare with the current state of this branch before deciding to go with it. In short, the refactor I propose also affects createTypedFunction -- basically I delay the introduction of conversions, since they don't do anything until it's time to construct the implementation, and keep the signature information just in a map until then, so that resolveReferences just has a signature map to deal with (not also an array of signatures). That seemed to me to simplify resolveReferences considerably. Anyhow, let me know if there's anything else I can do to assist with this, such as file a pull request to this branch or adjust the proposed approach to refactoring.

Reply:
Thanks for keeping your refactoring in a separate branch, that indeed makes it easy for me to see what you did exactly. I'm really happy with this refactoring, it makes sense to move introducing conversions to the end!

I'll merge your refactor into the `feat/no_binding_self2` branch and do some final tuning (there is no transpiler in place so to keep IE11 happy we shouldn't use modern features, and I prefer `clearResolutions` and `resolveReferences` to not mutate the input arguments but create a new map)

Reply:
Will now merge this PR, and address making things work on IE11 separately (there where also some other recent changes that are not IE11 compatible 😅 )

Reply:
Probably my fault. Sorry I don't have a clear idea in my head exactly what are IE11 transgressions... Is there a cheat sheet somewhere of "things not to do for IE11"?